contracts/src/libraries/LPMath.sol

```diff
@@ -127,6 +127,113 @@ library LPMath {
+    function calculateLongExposure(
+        uint256 _longExposure,
+        int256 _before,
+        int256 _after
+    ) internal pure returns (uint256) {
+        // The global long exposure is the sum of the non-netted longs in each
+        // checkpoint. To update this value, we subtract the current value
+        // (`_before.max(0)`) and add the new value (`_after.max(0)`).
+        int256 delta = FixedPointMath.max(_after, 0) -
+            FixedPointMath.max(_before, 0);
+        if (delta > 0) {
+            _longExposure += uint256(delta);
+        } else if (delta < 0) {
+            _longExposure -= uint256(-delta);
+        }
+
+        return _longExposure;
+    }
+
+    /// @dev Verifies that the price discovery process is valid.
+    /// @param _shareReserves The share reserves.
+    /// @param _shareAdjustment The share adjustment.
+    /// @param _bondReserves The bond reserves.
+    /// @param _minimumShareReserves The minimum share reserves.
+    /// @param _initialVaultSharePrice The initial vault share price.
+    /// @param _vaultSharePrice The vault share price.
+    /// @param _timeStretch The time stretch.
+    /// @param _checkpointExposure The checkpoint exposure.
+    /// @param _longExposure The global long exposure.
+    /// @return A flag indicating if the price discovery is valid.
+    function verifyPriceDiscovery(
+        uint256 _shareReserves,
+        int256 _shareAdjustment,
+        uint256 _bondReserves,
+        uint256 _minimumShareReserves,
+        uint256 _initialVaultSharePrice,
+        uint256 _vaultSharePrice,
+        uint256 _timeStretch,
+        int256 _checkpointExposure,
+        uint256 _longExposure
+    ) external pure returns (bool) {
+        // Calculate the share payment and bond proceeds of opening the
+        // largest possible long on the YieldSpace curve. This does not
+        // include fees.
+        (uint256 effectiveShareReserves, bool success) = HyperdriveMath
+            .calculateEffectiveShareReservesSafe(
+                _shareReserves,
+                _shareAdjustment
+            );
+        if (!success) {
+            return false;
+        }
+        uint256 maxSharePayment;
+        (maxSharePayment, success) = YieldSpaceMath.calculateMaxBuySharesInSafe(
+            effectiveShareReserves,
+            _bondReserves,
+            ONE - _timeStretch,
+            _vaultSharePrice,
+            _initialVaultSharePrice
+        );
+
+        uint256 maxBondProceeds;
+        (maxBondProceeds, success) = YieldSpaceMath.calculateMaxBuyBondsOutSafe(
+            effectiveShareReserves,
+            _bondReserves,
+            ONE - _timeStretch,
+            _vaultSharePrice,
+            _initialVaultSharePrice
+        );
+
+        // If one fails they should both be zero.
+        if (maxSharePayment == 0 || maxBondProceeds == 0) {
+            maxSharePayment = 0;
+            maxBondProceeds = 0;
+        }
```

Comment:
We could do something more efficient in the case of adding liquidity at 0% APR, but i do think they should be able to add liquidity at 0% if they want to.

Reply:
This makes me a bit nervous. Is the only time these calculations will fail when the APR is zero?

Reply:
I suggest playing with it some. let me ask a different question: why would we not allow people to add liquidity if we are solvent?

Reply:
Two things:

1. we’re throwing away the result of the failure flag which seems unsafe to me
2. We’re assuming that if one is zero, they should both be zero. I don’t know why that’s a good assumption (especially considering (1))

Reply:
1) my assumption here is the these two failure calcs are supposed to recreate the state that we would be in if we performed a max long. if one fails, then we are already in the state where a long can't be opened. that, in combination with the solvency constraint should still be valid. if we change it to trigger the circuit breaker on any of these conditions there are a bunch of perfectly reasonable and solvent states (if we believe our solvency equation) where people just can't add liquidity. it's not a good user experience bc if you revert and look at the logs you will see a bunch of cases where valid states don't allow you to add liquidity.

2) I understand; however, it seemed weird to test solvency by only applying one or the other. there are plenty of cases where one passes and the other fails. since only applying one doesn't capture the spirit of, "what's the state solvency if a max long is opened?" it seemed weird to only apply one. that being said, I could be convinced to change it to where we do this - from my pov it makes the solvency calc a little more conservative which is totally fine.

tl;Dr, I'm open to not doing (2). this is how I had it originally - I just wanted to explain the reasoning. regarding (1), I suggest we talk this out

Reply:
it seems like throwing out the failure flag is fine in this case, since checking if either is 0 covers the cases where it failed and where maxbuy is 0.... 

are there any cases where one would return 0 (without failing) and the other would return something nonzero?

Reply:
I don’t think the case where it failed and the max buy is zero is the issue. The issue is where it failed, the max buy is nonzero, and this implementation treats it like the max buy is zero. In this case, we’re not checking the solvency at the true max buy. That said, I haven’t looked through the max buy implementation to see if there are cases where it would actually fail in this way, but that’s what I would be concerned about.

Reply:
I reviewed this, and what you did makes sense to me @jrhea. I added a comment or two to explain the reasoning, but this part of the implementation was basically left unchanged.